x/participationrewards/keeper/submodule_umee.go

```diff
@@ -0,0 +1,195 @@
+		isupply, err := types.UnmarshalProtocolData(types.ProtocolDataTypeUmeeUTokenSupply, data.Data)
+		if err != nil {
+			return false
+		}
+		supply, _ := isupply.(*types.UmeeUTokenSupplyProtocolData)
+
+		// update utoken supply
+		k.IcqKeeper.MakeRequest(
+			ctx,
+			connectionData.ConnectionID,
+			connectionData.ChainID,
+			"store/leverage/key",
+			umeetypes.KeyUTokenSupply(supply.Denom),
+			sdk.NewInt(-1),
+			types.ModuleName,
+			UmeeUTokenSupplyUpdateCallbackID,
+			0,
+		) // query utoken supply
+
+		return false
+	})
+
+	// TODO: check module spendable coins retrieval
+	// assuming that module account is not a vesting account so there
+	// will be no locked coins to subtract from the total balance
+	k.IteratePrefixedProtocolDatas(ctx, types.GetPrefixProtocolDataKey(types.ProtocolDataTypeUmeeLeverageModuleBalance), func(idx int64, _ []byte, data types.ProtocolData) bool {
+		ibalance, err := types.UnmarshalProtocolData(types.ProtocolDataTypeUmeeLeverageModuleBalance, data.Data)
+		if err != nil {
+			return false
+		}
+		balance, _ := ibalance.(*types.UmeeLeverageModuleBalanceProtocolData)
+		accountPrefix := banktypes.CreateAccountBalancesPrefix(authtypes.NewModuleAddress(umeetypes.LeverageModuleName))
+
+		// update leverage module balance
+		k.IcqKeeper.MakeRequest(
+			ctx,
+			connectionData.ConnectionID,
+			connectionData.ChainID,
+			icstypes.BankStoreKey,
+			append(accountPrefix, []byte(balance.Denom)...),
+			sdk.NewInt(-1),
+			types.ModuleName,
+			UmeeLeverageModuleBalanceUpdateCallbackID,
+			0,
+		) // query leverage module balance
+
+		return false
+	})
+	// umee total borrowed update
+	k.IteratePrefixedProtocolDatas(ctx, types.GetPrefixProtocolDataKey(types.ProtocolDataTypeUmeeTotalBorrows), func(idx int64, _ []byte, data types.ProtocolData) bool {
+		iborrows, err := types.UnmarshalProtocolData(types.ProtocolDataTypeUmeeTotalBorrows, data.Data)
+		if err != nil {
+			return false
+		}
+		borrows, _ := iborrows.(*types.UmeeTotalBorrowsProtocolData)
+
+		// update total borrows for a denom
+		k.IcqKeeper.MakeRequest(
+			ctx,
+			connectionData.ConnectionID,
+			connectionData.ChainID,
+			"store/leverage/key",
+			umeetypes.KeyAdjustedTotalBorrow(borrows.Denom),
+			sdk.NewInt(-1),
+			types.ModuleName,
+			UmeeTotalBorrowsUpdateCallbackID,
+			0,
+		) // query leverage module balance
+
+		return false
+	})
+}
+
+func (u UmeeModule) ValidateClaim(ctx sdk.Context, k *Keeper, msg *types.MsgSubmitClaim) (uint64, error) {
+	amount := uint64(0)
+	_, addr, err := bech32.DecodeAndConvert(msg.UserAddress)
+
+	for _, proof := range msg.Proofs {
+		// determine denoms from keys
+		if proof.Data == nil {
+			continue
+		}
+
+		udenom, err := utils.DenomFromRequestKey(proof.Key, addr)
+		if err != nil {
+			return 0, err
+		}
+		uToken, err := bankkeeper.UnmarshalBalanceCompat(k.cdc, proof.Data, udenom)
```

Comment:
Does the leverage module store the tokens in the same way as the bank token? 

e.g an `sdk.Int` stored under the key `prefix + address + denom`

Reply:
the balance comes from the bankkeeper by querying all spendable coins after converting the module name to an address, I think it will be the same but for verification can look at x/bank/keeper/view.go/IterateAccountBalances()

Reply:
Sure, this works for the bank balance query, but I meant the leverage module proof.

https://github.com/umee-network/umee/blob/af7097083d6acf76cdc8c35c2d6d5e51e46368ef/x/leverage/types/keys.go#L52

seems to suggest to get the denom for the leverage proofs (switch on proof.Type?) you need to decode 

`collateralPrefix | lengthprefixed(addr) | denom | 0x00 for null-termination` to fetch the denom, and then the value is just an sdk.Int.

The above will work for balances, but not for the second proof, for the leverage module.

Reply:
looking at it, `udenom, err := utils.DenomFromRequestKey(proof.Key, addr)` on l179 _might_ work, depending on how Go treat null terminated strings. Technically it will return `u/uumee0x00` but I don't know if that will match `u/uumee`. Given that we know the denom is suffixed with a null byte, worst case we can drop the last byte in the event proof.Type is leverage.

Reply:
Would be good to add tests for ValidateClaim to cover both leverage and bank module responses.

Reply:
Yes the current key being used in xcclookup does not include the extra byte to prevent null termination either.
Adding the change there and handling the last byte as well